### PR TITLE
Bugfix/snmp engine

### PR DESCRIPTION
**Pull request by mjurczak**
Initial work to fix most critical issues described in SNMP agent implementation:
https://github.com/contiki-ng/contiki-ng/issues/1354
https://github.com/contiki-ng/contiki-ng/issues/1353
https://github.com/contiki-ng/contiki-ng/issues/1352
https://github.com/contiki-ng/contiki-ng/issues/1351

The proposed changes require testing and do not cover all out-of-bounds conditions that may occur when processing incoming SNMP requests. 

I think it would also be great if there was an automated method of functional validation of the SNMP agent implementation.

### CODE REVIEW

**Comment by Yagoor** [comment]
Hi mjurczak,
I implemented the SNMP protocol last year.

I believe you have some tests scenarios that you used to catch these bugs.
Could you add them to the unit test?
The unit test basically only validates the protocol, meaning that it makes sure it is compliant, but it doesn't really stresses it.

**Comment by mjurczak** [comment]
Hi Yagoor,

Some sample inputs are provided in issue describtions. I will provide you with more samples that cause similar porcessing issues.

I will also try to figure out which issues can be verified by UTs. For instance - if the number of requested OIDs exceeds supported value, the parser should be expected to return an error I think.
On the other hand, writes and reads beyond buffer boundaries and certain class of stack overwrite issues may either pass unnoticed in UTs or cause UT framework crash.

Most of the described issues were discovered by means of fuzzing. It would be my pleasure to post a sample base app and a short description on how to run fuzzer on the SNMP agent.

Kind regards.

**Review by Yagoor — APPROVED** [review]
Hi @g-oikonomou.
I took I look into his changes and they all seems pretty neat.

However in order to prevent these issues and making sure they remain fixed I suggest adding the test cases described in the issues he linked to this PR.
Unfortunately, I've been pretty busy at work these days. 

@mjurczak, would you be able to create the unit tests? I have an idea on how to do it. 
If you don't have time I believe a good approach would be to merge this PR without the unit tests and create an issue and assign it to me and in the future I'll create the unit tests and maybe others. Having the issue will keep this under my radar and I won't forget it.

**Comment by g-oikonomou** [comment]
If this is stand-alone and can be merged without further work I can for sure merge this. Thanks for reviewing!

**Comment by g-oikonomou** [comment]
@mjurczak @Yagoor are we ready here? Can I merge this?

**Comment by Yagoor** [comment]
HI @g-oikonomou. I'm working on more improvements in the protocol implementation.
After his points, I reviewed the code and notice more vulnerabilities (corner cases).

I see that you added it to version 4.6 milestones. I'll try to get it done for it.
I'm implementing several unit tests in the engine to make sure it is not only compliant with the protocol but that is also handles corner cases.

File: os/net/app-layer/snmp/snmp-ber.c [hunk 1]
```diff
@@ -128,6 +128,11 @@ snmp_ber_encode_null(unsigned char *out, uint32_t *out_len, uint8_t type)
 unsigned char *
 snmp_ber_decode_type(unsigned char *buff, uint32_t *buff_len, uint8_t *type)
 {
+  if(*buff_len == 0)
+  {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-ber.c:132]: Style - opening brace in the same line with the if, like so:

`if(*buff_len == 0) {`

File: os/net/app-layer/snmp/snmp-ber.c [hunk 2]
```diff
@@ -137,6 +142,11 @@ snmp_ber_decode_type(unsigned char *buff, uint32_t *buff_len, uint8_t *type)
 unsigned char *
 snmp_ber_decode_length(unsigned char *buff, uint32_t *buff_len, uint8_t *length)
 {
+  if(*buff_len == 0)
+  {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-ber.c:146]: Ditto

File: os/net/app-layer/snmp/snmp-ber.c [hunk 3]
```diff
@@ -160,14 +171,20 @@ snmp_ber_decode_integer(unsigned char *buf, uint32_t *buff_len, uint32_t *num)
...
+     len > 4) {
     /*
      * Sanity check
      * It will not fit in the uint32_t
      */
     return NULL;
   }
 
+  if(*buff_len < len)
+  {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-ber.c:184]: Ditto

File: os/net/app-layer/snmp/snmp-ber.c [hunk 4]
```diff
@@ -150,7 +160,8 @@ snmp_ber_decode_integer(unsigned char *buf, uint32_t *buff_len, uint32_t *num)
 
   buf = snmp_ber_decode_type(buf, buff_len, &type);
 
-  if(type != BER_DATA_TYPE_INTEGER) {
+  if(buf == NULL ||
+     type != BER_DATA_TYPE_INTEGER) {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-ber.c:164]: You _could_ keep those in the same line, but you don't have to if you cannot be bothered

File: os/net/app-layer/snmp/snmp-ber.c [hunk 5]
```diff
@@ -196,14 +214,20 @@ snmp_ber_decode_unsigned_integer(unsigned char *buf, uint32_t *buff_len, uint8_t
...
+     len > 4) {
     /*
      * Sanity check
      * It will not fit in the uint32_t
      */
     return NULL;
   }
 
+  if(*buff_len < len)
+  {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-ber.c:227]: Ditto

File: os/net/app-layer/snmp/snmp-ber.c [hunk 6]
```diff
@@ -222,7 +246,8 @@ snmp_ber_decode_string_len_buffer(unsigned char *buf, uint32_t *buff_len, const
 
   buf = snmp_ber_decode_type(buf, buff_len, &type);
 
-  if(type != BER_DATA_TYPE_OCTET_STRING) {
+  if(buf == NULL ||
+    type != BER_DATA_TYPE_OCTET_STRING) {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-ber.c:250]: Same line - optional

File: os/net/app-layer/snmp/snmp-ber.c [hunk 7]
```diff
@@ -196,14 +214,20 @@ snmp_ber_decode_unsigned_integer(unsigned char *buf, uint32_t *buff_len, uint8_t
 
   buf = snmp_ber_decode_length(buf, buff_len, &len);
 
-  if(len > 4) {
+  if(buf == NULL ||
+     len > 4) {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-ber.c:218]: Same line - optional

File: os/net/app-layer/snmp/snmp-ber.c [hunk 8]
```diff
@@ -186,7 +203,8 @@ snmp_ber_decode_unsigned_integer(unsigned char *buf, uint32_t *buff_len, uint8_t
 
   buf = snmp_ber_decode_type(buf, buff_len, &type);
 
-  if(type != expected_type) {
+  if(buf == NULL ||
+     type != expected_type) {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-ber.c:207]: Same line - optional

File: os/net/app-layer/snmp/snmp-ber.c [hunk 9]
```diff
@@ -160,14 +171,20 @@ snmp_ber_decode_integer(unsigned char *buf, uint32_t *buff_len, uint32_t *num)
 
   buf = snmp_ber_decode_length(buf, buff_len, &len);
 
-  if(len > 4) {
+  if(buf == NULL ||
+     len > 4) {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-ber.c:175]: Same line - optional

File: os/net/app-layer/snmp/snmp-message.c [hunk 10]
```diff
@@ -224,6 +224,10 @@ snmp_message_decode(uint8_t *buf, uint32_t buf_len, snmp_header_t *header,
   }
 
   for(i = 0; buf_len > 0; ++i) {
+    if (i >= *varbind_num)
+    {
```
Comment (g-oikonomou) [inline os/net/app-layer/snmp/snmp-message.c:228]: Ditto

**Review by g-oikonomou — CHANGES_REQUESTED** [review]
Hi @mjurczak thank you for this, I finally managed to make some time to review. Looks good to me. Please if you can fix the code style as per the comments, then I am keen to merge this one. Let me know if you have no time and I'll apply the fixes.

File: os/net/app-layer/snmp/snmp-message.c (hunk 10 above)
Comment (g-oikonomou) [reply os/net/app-layer/snmp/snmp-message.c:228]: Also no space between if and opening paren

**Review by g-oikonomou — APPROVED** [review]

**Comment by g-oikonomou** [comment]
Style errors fixed in #1414 instead of blocking this one
